stan/math/prim/mat/meta/promote_vector.hpp

```diff
@@ -0,0 +1,42 @@
+#ifndef STAN_MATH_PRIM_MAT_META_PROMOTE_VECTOR_HPP
+#define STAN_MATH_PRIM_MAT_META_PROMOTE_VECTOR_HPP
+
+#include <boost/utility/enable_if.hpp>
+#include <boost/math/tools/promotion.hpp>
+#include <stan/math/prim/scal/meta/is_vector.hpp>
+
+namespace stan {
+  /**
+   * promote_vector is a metaprogram that returns the type of the first
+   * vector-like type in the arguments list.
+   * If there are no vector-like types, it returns the first scalar type in its
+   * argument list promoted according to the boost type promotion rules.
+   *
+   * @tparam T1 First argument
+   * @tparam T2 Second argument
+   * @tparam Enable (used internally, don't specify)
+   */
+
+  template <typename T1, typename T2, typename Enable = void>
+  struct promote_vector {
+    typedef typename boost::math::tools::promote_args<T1>::type type;
+  };
+
+  template <typename T1, typename T2>
+  struct promote_vector<T1, T2, typename boost::enable_if_c<
+                                  is_vector<T1>::value &&
+                                  !is_vector<T2>::value
+                                  >::type > {
+    typedef T1 type;
+  };
+
+  template <typename T1, typename T2>
+  struct promote_vector<T1, T2, typename boost::enable_if_c<
+                                  !is_vector<T1>::value &&
+                                  is_vector<T2>::value
+                                  >::type > {
+    typedef T2 type;
```

Comment:
You're right. I had a look at the distributions that exist. I think the goal should be to support only as much vectorization as the sampling statements do.

So ```multinomial_rng``` is the one that returns ```int[]```. It already matches the vectory-ness of the sampling statement, so that's good enough.

```multi_normal_rng``` (which returns a vector) has sampling signatures that allow lists of vectors to be passed in, so I think the rng should allow the same at the output.

```dirichlet```, however, doesn't have sampling signatures that allow lists of simplexes to be passed in. So I don't think it should be vectorized.

The matrix distributions aren't vectorized either. This leaves rngs that do:

* int
* real
* vector

I think the only return type that makes sense for ```multi_normal``` would be arrays of vectors (I don't like the idea of promoting to matrices).

To be consistent, all the other rngs would return arrays as well?

I think it makes sense to allow arbitrary mixes of arrays, vectors, and row_vectors on the inputs, just like the samplers, but exclusively return arrays of things.

Does this logic sound reasonable?